### PR TITLE
added support for Pragma

### PR DESCRIPTION
**Pull request by cod3monk**
Is there any chance to include (optional) support for #pragma in the AST? I know it is not strictly C99, since it is supposed to be handled before and not through the AST.

Dirty patch included, just as an example.

### CODE REVIEW

**Comment by eliben** [comment]
Yes, I'm fine with including pragmas in the AST, as long as the change is conservative - i.e. doesn't break other code and if we encounter pragmas we can tokenize but not parse yet it shouldn't fail (similarly to the way it's done now).

**Comment by cod3monk** [comment]
I added and modified the unit test accordingly. Any other requests? 
Do you think the additional token will break any code?

File: pycparser/c_ast.py [hunk 1]
```diff
@@ -795,3 +795,15 @@ def children(self):
 
     attr_names = ()
 
+class Pragma(Node):
```
Comment (eliben) [inline pycparser/c_ast.py]: You are not supposed to modify this file directly. It's auto generated

File: pycparser/c_lexer.py [hunk 2]
```diff
@@ -171,7 +171,9 @@ def _make_tok_location(self, token):
         'ELLIPSIS',
 
         # pre-processor
-        'PPHASH',      # '#'
+        'PPHASH',       # '#'
+        'PPPRAGMA',     # '#pragma'
```
Comment (eliben) [inline pycparser/c_lexer.py]: PPPRAGMA is for 'pragma', not '#pragma', right?

File: pycparser/c_lexer.py [hunk 3]
```diff
@@ -304,15 +305,18 @@ def t_pppragma_NEWLINE(self, t):
...
+    t_pppragma_ignore = ' \t'
 
+    def t_pppragma_STR(self, t):
+        '.+'
+        t.type = 'PPPRAGMASTR'
+        return t
+    
     @TOKEN(identifier)
-    def t_pppragma_ID(self, t): pass
+    def t_pppragma_ID(self, t):
```
Comment (eliben) [inline pycparser/c_lexer.py]: Why is this still ignored?

File: pycparser/c_ast.py (hunk 1 above)
Comment (cod3monk) [inline pycparser/c_ast.py]: Yes, I forgot to add the changed _c_ast.cfg file to the commit. Added in 95ee04fffc2530a7f9966b517d5d9173e91e5b0e

File: pycparser/c_parser.py [hunk 4]
```diff
@@ -547,9 +547,17 @@ def p_external_declaration_4(self, p):
 
     def p_pp_directive(self, p):
         """ pp_directive  : PPHASH
+                          | PPPRAGMA
```
Comment (eliben) [inline pycparser/c_parser.py]: I'd have a separate parsing rule for pragma now that it's supported, leaving the other one as it was

File: tests/test_c_lexer.py [hunk 5]
```diff
@@ -309,28 +309,52 @@ def test_preprocessor_line_funny(self):
...
+        #\tpragma {pack: 2, smack: 3}
         #pragma <includeme.h> "nowit.h"
         #pragma "string"
         #pragma somestring="some_other_string"
         #pragma id 124124 and numbers 0235495
         59
         '''
-
-        # Check that pragmas are ignored but the line number advances
+        # Check that pragmas are tokanized, including trailing string
```
Comment (eliben) [inline tests/test_c_lexer.py]: s/tokanized/tokenized/

File: pycparser/c_lexer.py (hunk 3 above)
Comment (cod3monk) [inline pycparser/c_lexer.py]: Is unused and can safely be removed. Should I remove it?

Comment (eliben) [inline pycparser/c_lexer.py]: Yes, go ahead

**Comment by cod3monk** [comment]
Anything else missing?

**Comment by eliben** [comment]
Looks good. Please squash into a single commit and put as much info as possible into the commit message; then it's good to go.

**Comment by eliben** [comment]
Thanks!

**Pull request merged**